opm/autodiff/AquiferCarterTracy.hpp

```diff
@@ -0,0 +1,505 @@
+                W_flux_ = 0.;
+
+                // We next get our connections to the aquifer and initialize these quantities using the initialize_connections function
+                initialize_connections(connection);
+
+                // pa0_ is the initial aquifer water pressure. Must be calculated from equilibrium if left default,
+                // or we get the information from the deck Hacked to make it at 45e6 Pa
+                calculate_reservoir_equilibrium();
+                
+                pressure_previous_.resize(cell_idx_.size(), 0.);
+                pressure_current_.resize(cell_idx_.size(), 0.);
+
+                rhow_.resize(cell_idx_.size(), 998.0); 
+                Qai_.resize(cell_idx_.size(), 0.);
+
+                polynomial_fit(aqutab_td_, aqutab_pi_, coeff_, 2, true);
+            }
+
+            inline void get_current_Pressure_cell(std::vector<Scalar>& pressure_water, const int idx, const IntensiveQuantities& intQuants)
+            {
+                const auto& fs = intQuants.fluidState();
+                pressure_water[idx] = fs.pressure(FluidSystem::waterPhaseIdx).value();
+            }
+
+            inline void get_current_density_cell(std::vector<Scalar>& rho_water, const int idx, const IntensiveQuantities& intQuants)
+            {
+                const auto& fs = intQuants.fluidState();
+                rho_water[idx] = fs.density(FluidSystem::waterPhaseIdx).value();
+            }
+
+            inline Scalar dpai(int idx)
+            {
+                Scalar dp = pa0_ - rhow_[idx]*gravity_*(cell_depth_[idx] - d0_) - pressure_previous_[idx];
+                return dp;
+            }
+
+            inline void calculate_a_b_constants(Scalar& a, Scalar& b, const int idx, const SimulatorTimerInterface& timer)
+            {
+                // This function implements Eqs 5.8 and 5.9 of the EclipseTechnicalDescription
+                Scalar beta = aquifer_influx_constant();
+                Scalar Tc = time_constant();
+                Scalar td_plus_dt = (timer.currentStepLength() + timer.simulationTimeElapsed()) / Tc;
+                Scalar td = timer.simulationTimeElapsed() / Tc;
+                Scalar PItdprime = coeff_[1] + 2.0*coeff_[2]*(td_plus_dt);
+                Scalar PItd = coeff_[0] + coeff_[1]*td_plus_dt + coeff_[2]*td_plus_dt*td_plus_dt;
+                a = 1.0/Tc * ( (beta * dpai(idx)) - (W_flux_ * PItdprime) ) / ( PItd - td*PItdprime );
+                b = beta / Tc / ( PItd - td*PItdprime);
+            }
+
+            inline void calculate_inflow_rate(int idx, const SimulatorTimerInterface& timer)
+            {
+                Scalar a, b;
+                calculate_a_b_constants(a,b,idx,timer);
+                // This function implements Eq 5.7 of the EclipseTechnicalDescription
+                Qai_[idx] = area_fraction(idx)*( a - b * ( pressure_current_[idx] - pressure_previous_[idx] ) );
```

Comment:
Based on this formulation, probably the derivatives of pressure_current_ should be kept.